libbeat/outputs/elasticsearch/client_proxy_test.go

```diff
@@ -0,0 +1,246 @@
+	// The child process always runs only the TestClientPing test, which pings
+	// the server at TEST_SERVER_URL and then terminates.
+	cmd := exec.Command(os.Args[0], "-test.run=TestClientPing")
+	cmd.Env = append(append(os.Environ(),
+		"TEST_START_CLIENT=1"),
+		env...)
+	return cmd.Run()
+}
+
+func doClientPing(t *testing.T) {
+	serverURL := os.Getenv("TEST_SERVER_URL")
+	proxy := os.Getenv("TEST_PROXY_URL")
+	headerURL := os.Getenv("TEST_HEADER_URL")
+	clientSettings := ClientSettings{
+		URL:     serverURL,
+		Index:   outil.MakeSelector(outil.ConstSelectorExpr("test")),
+		Headers: map[string]string{"X-Test-URL": serverURL},
+	}
+	if proxy != "" {
+		proxyURL, err := url.Parse(proxy)
+		assert.NoError(t, err)
+		clientSettings.Proxy = proxyURL
+	}
+	if headerURL != "" {
+		// Some tests point at an intentionally invalid server because golang's
+		// helpers never return a proxy for a request to the local host (see
+		// TestEnvironmentProxy), and in that case we still want to record the real
+		// server URL in the headers for verification.
+		clientSettings.Headers["X-Test-URL"] = headerURL
+	}
+	client, err := NewClient(clientSettings, nil)
+	assert.NoError(t, err)
+
+	// This ping won't succeed; we aren't testing end-to-end communication
+	// (which would require a lot more setup work), we just want to make sure
+	// the client is pointed at the right server or proxy.
+	client.Ping()
+}
+
+// testListeners is a wrapper that handles starting up http listeners
+// representing an elastic server and an intermediate proxy, confirming that
+// requests target the expected (server) URL, and counting how many requests
+// arrive at each endpoint.
+type testListeners struct {
+	server *httptest.Server
+	proxy  *httptest.Server
+
+	// This mutex must be held in order to modify the structure.
+	// Using a mutex for servers that are expected to receive ~1 request in
+	// their lifetime might be pedantic, but I really dislike race conditions.
+	mutex sync.Mutex
+
+	serverRequestCount int // Requests directly to the server
+	proxyRequestCount  int // Requests via the proxy
+
+	errors []string
+}
+
+// start creates and starts the server and proxy listeners.
+func (tl *testListeners) start(t *testing.T) {
+	tl.mutex.Lock()
+	defer tl.mutex.Unlock()
+	tl.server = httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			tl.mutex.Lock()
+			defer tl.mutex.Unlock()
+			tl._assertEqual("server handler", tl.server.URL, r.Header.Get("X-Test-URL"))
+			fmt.Fprintln(w, "Hello, client")
+			tl.serverRequestCount++
+		}))
+	tl.proxy = httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			tl.mutex.Lock()
+			defer tl.mutex.Unlock()
+			tl._assertEqual("proxy handler", tl.server.URL, r.Header.Get("X-Test-URL"))
+			fmt.Fprintln(w, "Hello, client")
+			tl.proxyRequestCount++
+		}))
+}
+
+func (tl *testListeners) close() {
+	tl.server.Close()
+	tl.proxy.Close()
+}
```

Comment:
I noticed we have a common pattern here: 

```
tl := testListeners()
tl.start()
defer tl.close()
```

If we don't need access to server or proxy we could also provide a setup + teardown routine (and separate test state from servers):

```
type state struct {
    mutex sync.Mutex

	serverRequestCount int // Requests directly to the server
	proxyRequestCount  int // Requests via the proxy
}

func TestX(t *testing.T) {
   st, teardown := setupServers(t)
   defer teardown()

   ...
}

func setupServers(t *testing.T) (*state, func()) {
  st := &state{}
  server := http.Server( ... )
  proxy := http.Server( ... )
  return st, func() {
    server.Close()
    proxy.Close()
  }
}
```

No need to change it. Either pattern works for me.

Reply:
i like this pattern a lot, i have switched to it (though i also had to add the endpoint URLs to the state)